File: AdvancedGraphs/reconstruct_itinerary.py
class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        # U will need to run DFS on JFK since we will start from there
        # we need to build a adj list in which we can use it to map out the itinerary
        # First we sort the input lists and when we add the tickets in they will be in order
        # we know its done when len(ans) = len(tickets) + 1
        # To solve deadends, when we meet a deadend, we add it back to the adj[key]
        # Time complexity: Best O(V+E), Worse O(V+E^2), Space complexity: O(E)
        ans = ["JFK"] # start with JFK first
        adj = {i:[] for i,v in tickets}
        tickets.sort()
        for i,v in tickets:
            adj[i].append(v)
        print(adj)
        def dfs(source):
            if len(ans) == len(tickets) +1:
                return True
            if source not in adj: # means one way trip into dest node with no way back
                return False
            temp = list(adj[source])
            for i, v in enumerate(temp):
                adj[source].pop(i) # pop the index of the node u visiting next
                ans.append(v)
                if dfs(v):
                    return True
                adj[source].insert(i,v)
                ans.pop()
                print(adj)
            return False

        dfs("JFK")
        return ans

File: AdvancedGraphs/cpp/reconstruct_itinerary.cpp
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        vector<string> ans;
        unordered_map<string, vector<string>> adj; // Use unordered_map with string keys
        sort(tickets.begin(), tickets.end());

        for (const vector<string>& ticket : tickets) {
            string from = ticket[0];
            string to = ticket[1];
            adj[from].push_back(to);
        }

        dfs("JFK", ans, adj, tickets);
        return ans;
    }

    bool dfs(const string& source, vector<string>& ans, unordered_map<string, vector<string>>& adj, vector<vector<string>>& tickets) {
        ans.push_back(source);

        if (ans.size() == tickets.size() + 1) {
            return true;
        }

        if (adj.find(source) == adj.end()) {
            ans.pop_back();
            return false;
        }

        vector<string> temp = adj[source];

        for (size_t i = 0; i < temp.size(); ++i) {
            string v = temp[i];
            adj[source].erase(adj[source].begin() + i);
            if (dfs(v, ans, adj, tickets)) {
                return true;
            }
            adj[source].insert(adj[source].begin() + i, v);
        }

        ans.pop_back();
        return false;
    }
};
